Convert YoloV2 flags to EIP-2315 and EIP-2929

ziogaschr opened the pull request.

ziogaschr committed:
parmas/types: add support for EIP2315 and EIP2929

ziogaschr committed:
params/types/coregeth: fix wrong json property `eip2315FBlock` name used for `EIP2537FBlock`

ziogaschr committed:
cmd/evm/internal/t8ntool, core, core/vm, eth: convert YoloV2 enablement checks to the corresponding EIPs

ziogaschr committed:
temp fix for —yolov2 test

ziogaschr committed:
params/types/parity: Add EIP2315Transition and EIP2929Transition for parity_configurator.go

ziogaschr committed:
params/types/multigeth: Add EIP2315Transition and EIP2929Transition

not yet pulled from upstream to multigeth

ziogaschr commented on core/vm/runtime/runtime.go:
```diff
@@ -113,17 +113,14 @@ func Execute(code, input []byte, cfg *Config) ([]byte, *state.StateDB, error) {
+		cfg.State.AddAddressToAccessList(cfg.Origin)
+		cfg.State.AddAddressToAccessList(address)
+		for addr := range vm.PrecompiledContractsForConfig(cfg.ChainConfig, vmenv.BlockNumber) {
+			cfg.State.AddAddressToAccessList(addr)
```
The duplicated line here, is already being tracked at [go-ethereum](https://github.com/ethereum/go-ethereum/pull/21956)

ziogaschr commented on params/config.go:
```diff
@@ -30,8 +30,7 @@ var (
 	GoerliGenesisHash  = common.HexToHash("0xbf7e331f7f7c1dd2e05159666b3bf8bc7a8a3a9eb1d518969eab529dd9b88c1a")
-	// TODO: update with yolov2 values
-	YoloV2GenesisHash = common.HexToHash("0x498a7239036dd2cd09e2bb8a80922b78632017958c332b42044c250d603a8a3e")
+	YoloV2GenesisHash  = common.HexToHash("0xc3fd235071f24f93865b0850bd2a2119b30f7224d18a0e34c7bbf549ad7e3d36") // TODO: check if yolov2 is correct
```
At go-ethereum they left this on yolov1 until they get the new value. Though we have a testcase that fails. So I temporarily used the output of the test to make it happy

holiman replied:
Not really, we just chose to use the exact same genesis (alloc + time but not chainId) for yolov2 as yolov1.

ziogaschr replied:
Thanks @holiman for jumping in, this clears up everything. I will fix anything needed to make the test pass then.

ziogaschr commented on params/types/coregeth/chain_config.go:
```diff
@@ -163,7 +163,7 @@ type CoreGethChainConfig struct {
 
 	// https://github.com/ethereum/EIPs/pull/2537: BLS12-381 curve operations
-	EIP2537FBlock *big.Int `json:"eip2315FBlock,omitempty"`
+	EIP2537FBlock *big.Int `json:"eip2537FBlock,omitempty"`
```
This seems to be a mistype which I fixed. Correct me if it was intentionally

meowsbits replied:
You are correct, it was my mistake

ziogaschr commented on params/types/parity/parity.go:
```diff
@@ -106,8 +106,10 @@ type ParityChainSpec struct {
+		EIP2315Transition         *ParityU64 `json:"eip2315Transition,omitempty"`
 		EIP2537Transition         *ParityU64 `json:"eip2537Transition,omitempty"`
 		EIP1706Transition         *ParityU64 `json:"-"` // FIXME, when and if i'm implemented in Parity
+		EIP2929Transition         *ParityU64 `json:"-"` // FIXME, when and if i'm implemented in Parity
```
We have a test case on forks (TestEquivalent_Features) that fails on this transition. The thing is that OpenEthereum is not going to implement this EIP as they are out of resources at this moment. 

What is the best way to handle this test case @meowsbits?

meowsbits replied:
I'm guessing `func (spec *ParityChainSpec) SetEIP2929Transition(n *uint64) error {` needs to actually set `EIP2929Transition`, in the same way as EIP2315 is handled for Parity.

meowsbits replied:
Resolved with https://github.com/etclabscore/core-geth/pull/246/commits/f3aff0ec9c2ec7342973e19f739c881c226bd867

ziogaschr replied:
This resolves our test but I though of leaving it unsupported based on the comment [here](https://docs.google.com/spreadsheets/d/1BomvS0hjc88eTfx1b8Ufa6KYS3vMEb2c8TQ5HJWx2lc/edit#gid=408811124) saying that they are not going to implement this.

meowsbits replied:
Ah, roger that. I've reverted that commit and just modified the tests to skip Parity :: [Berlin|YOLO] tests. See https://github.com/etclabscore/core-geth/pull/246/commits/186e3cb59becdd812fa4c57da6ec3e4458e899dc and subsequent.

ziogaschr committed:
update YoloV2 chainId and change references to YoloV1 from code

ziogaschr commented on cmd/puppeth/wizard_genesis.go:
```diff
@@ -240,8 +240,8 @@ func (w *wizard) manageGenesis() {
 		fmt.Println()
-		fmt.Printf("Which block should YOLOv1 (EIP2537) come into effect? (default = %v)\n", w.conf.Genesis.Config.GetEIP2537Transition())
-		w.conf.Genesis.Config.SetEIP2537Transition(w.readDefaultUint64P(*w.conf.Genesis.Config.GetEIP2537Transition()))
+		fmt.Printf("Which block should YOLOv2 come into effect? (default = %v)\n", w.conf.Genesis.Config.GetEIP2929Transition())
```
@meowsbits there is not really an EIP for YoloV2 yet, so I used the EIP2929 as the flag for that

meowsbits replied:
Yeap, here's the EIP-as-fork-indicator scenario again. This might eventually warrant it's own logic somewhere (a fork :: indicator EIP function). Another important place where this happens is here, for example: https://github.com/etclabscore/core-geth/blob/master/core/vm/evmc.go#L302

ziogaschr replied:
Thanks for sharing the evmc link. Good to know that something like this exists.

meowsbits commented on eth/api_tracer.go:
```diff
@@ -597,22 +598,17 @@ func (api *PrivateDebugAPI) standardTraceBlockToFile(ctx context.Context, block
+		*chainConfigCopy = chainConfig
+		chainConfig = *chainConfigCopy
+		if eip2929 := config.Overrides.GetEIP2929Transition(); eip2929 != nil {
+			if err := chainConfig.SetEIP2929Transition(eip2929); err != nil {
```
So we have to use `config.Overrides.GetEIP2929Transition(); eip2929 != nil` to see if YoloV2 is activated (using an indicator EIP as a proxy for all associated EIPs for that fork.

But don't we have to do activations like `chainConfig.SetEIP2929Transition(eip2929)` for _all_ the EIPs in that fork (ie `chainConfig.SetEIP2315Transition(eip2929)`, etc.)?

ziogaschr replied:
Good one. Yes we have to handle all Yolo EIPs this way individually, I am not sure if we should use an indicator EIP to read the fork block number and use it to set all the EIPs on this block. The reason is that on some chain we might want to include only some of the EIPs and not all.

I will add `EIP2315` and `EIP2537` for now.

ziogaschr replied:
The overrides were actually used for producing an analysis for EIP-2929 and not being used in tests or anywhere else in the code for now. You can check [here](https://github.com/ethereum/go-ethereum/pull/21509#issue-477308710)

ziogaschr replied:
Resolved with https://github.com/etclabscore/core-geth/pull/246/commits/75af684928f0041d9c5df88776f3e956a508eda4

ziogaschr replied:
I am re-thinking of it @meowsbits .
1. What they wanted to test was actually only the EIP2929 implications, so the rest can be removed
2. We can even remove this block from here and consider it tested. Later on we can bring it back if we need to test similar E[C]IPs

meowsbits replied:
SGTM. Nice investigation!

ziogaschr replied:
@meowsbits so **leave the code as is** or go with **1** or **2**. I prefer 1.

meowsbits replied:
Yes, 1 SGTM.  In general I want to own and persist a superset of the tests from ethereum/go-ethereum (ie. never _only_ rely on them to have tested something, we need to own all the tests... and more ;).

ziogaschr replied:
Resolves with https://github.com/etclabscore/core-geth/pull/246/commits/26b814975e1c7cdf3afb9f81756c1c0f60446f21.
Which reverts the previous resolving commit :)

ziogaschr committed:
eth: breakdown YoloV2 overrides in api_tracer to corresponding EIPs

meowsbits commented on params/types/parity/parity_configurator.go:
```diff
@@ -985,6 +985,26 @@ func (spec *ParityChainSpec) SetEthashECIP1099Transition(n *uint64) error {
+	if n == nil {
+		return nil
+	}
+	return ctypes.ErrUnsupportedConfigFatal
```
Rel https://github.com/etclabscore/core-geth/pull/246/files#r536073142

meowsbits committed:
parity: implement ephemral EIP2929 for Parity config

Fixes ./integration/#TestEquivalent_Features

Resolves https://github.com/etclabscore/core-geth/pull/246/files#r535740801

Date: 2020-12-04 06:52:01-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr committed:
Revert "eth: breakdown YoloV2 overrides in api_tracer to corresponding EIPs"

This reverts commit 75af684928f0041d9c5df88776f3e956a508eda4.

meowsbits committed:
Revert "parity: implement ephemral EIP2929 for Parity config"

This reverts commit f3aff0ec9c2ec7342973e19f739c881c226bd867.

meowsbits committed:
integration: skip Parity conf feature equivalance checks for EIP2929 forks

EIP2929 is and will be unsupported by Parity, so
we cannot test chain config feature equivalence for
forks that demand that feature. Skip those tests.

Date: 2020-12-04 07:45:40-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Merge branch 'feat/support-eips-2315-2929-up' into feat/support-eips-2315-2929

meowsbits committed:
integration: fix bad logic

Date: 2020-12-04 07:57:11-06:00
Signed-off-by: meows <b5c6@protonmail.com>
